slither-function-filter tool

shortdoom opened the pull request:
reference: https://github.com/crytic/slither/issues/2233

`slither-function-filter` tool. for the time being it supports basic flag-based filtering (no way to specify details of search, only true/false)

two example inputs:

`slither-function-filter contract.sol --visibility external --modifiers --ext-calls` or

`slither-function-filter contract.sol --read-only` etc.

so, it's possible to:

_"Return all functions that are {visibility}, have/don't have internal OR external calls , have/don't have modifiers, are/aren't mutable (involve a potential state change)"_

all allowed flags:

```
--contract-name (str)
--visibility (str)
--modifiers (bool)
--ext-calls (bool)
--int-calls (bool)
--state-change (bool)
--read-only (bool)
```

every flag is optional, if none is specified, all contract functions will be returned

implementation is focused around output of function.get_summary() but filter_functions() is easy to extend.

output is formatted in logger and with coloring.

![image](https://github.com/crytic/slither/assets/56428630/79479ef9-e9fd-408e-805d-573ab32d35f0)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/crytic/slither?pullRequest=2238) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/crytic/slither?pullRequest=2238) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/crytic/slither?pullRequest=2238) it.</sub>

0xalpharush commented:
I had wanted something similar and made a note to myself:

>We should expose a tool for semantic grepping with the ability to target contract/ function names, filter by visibility/purity,
>Proposed CLI:
>
```
slither grep contracts/ [--contract|--function] [--pure,--nonpure] [--external-calls] [--unchecked] [--assembly]
```

I think that this is a decent start but I would imagine showing the [source](https://crytic.github.io/slither/slither/core/source_mapping/source_mapping.html#Source.content) and [line info](https://crytic.github.io/slither/slither/core/source_mapping/source_mapping.html#Source.to_detailed_str) so that it can be quickly examined or jumped to in the editor would be better. Wdyt?

shortdoom commented:
Some time passed from this PR but I remember that choosing exact `Contract.sol` as target to query is preferable over iterating the whole directory - avoids information overload in the cli output. 

So you won't need this part of the command:

`...contracts/ [--contract|--function]...` just `Contract.sol`

Current version was designed as _query contract to find function_. It's an interesting direction to also allow some grepping at the level of Function itself (has_require, has_loop, etc.).

I noticed you also included  `--unchecked` and `--assembly`, I would also add `--low_lvl_calls`. 

I also agree on adding one more string input flag to allow to _find a specific function_ by `full_name`. 

Showing source code of a Function is trickier UX-wise. Some functions will be long and cli output will get cluttered. Hence why I resorted to displaying structured info (it tends to be more or less the same length). Displaying a line reference is a great idea, most IDEs will just allow to jump automatically and we could avoid displaying source code in terminal.

To have a full _grep like tool_ we could allow to actually grep the function's source code without cluttering the cli though! 

`slither grep Contract.sol --match 'try:'` - that would find all functions using try/catch block ("try:" string) in the function source code.

I can ofc bring this PR home.

shortdoom commented:
Also, wouldn't it be cool to have it as PythonAPI? This is how I actually implemented it for my own use, as an external class taking a list of Function to filter, returning Function(s) matching. I needed to resort to writing an external query class because there's no API (and this PR also doesn't propose it in the current state)

shortdoom commented:
Added `lines` reference to the output (allows to Jump-to-line)

Added more filters:

```
--contains-asm (bool)
--low-lvl-calls (bool)
--full-name (str)
--in-source (str)
```

`full-name` allows user to search inside the function.full_name string (not an exact match)

`in-source` will search inside of the functions source code (not an exact match)

[slither-function-filter.webm](https://github.com/crytic/slither/assets/56428630/87cbe581-983c-4e69-9b91-65168d930e0a)


I think that covers most immediate needs we've seen mentioned by developers?
